go-controller/pkg/node/port_claim.go

```diff
@@ -0,0 +1,181 @@
+				for _, err := range errors {
+					klog.Error(err)
+				}
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			svc := obj.(*kapi.Service)
+			if errors := deleteServicePortClaim(svc); len(errors) > 0 {
+				for _, err := range errors {
+					klog.Error(err)
+				}
+			}
+		},
+	}, nil)
+	return err
+}
+
+func addServicePortClaim(svc *kapi.Service) []error {
+	return handleServicePortClaim(svc, true)
+}
+
+func deleteServicePortClaim(svc *kapi.Service) []error {
+	return handleServicePortClaim(svc, false)
+}
+
+func updateServicePortClaim(oldSvc, newSvc *kapi.Service) []error {
+	if reflect.DeepEqual(oldSvc.Spec.ExternalIPs, newSvc.Spec.ExternalIPs) && reflect.DeepEqual(oldSvc.Spec.Ports, newSvc.Spec.Ports) {
+		return nil
+	}
+	errors := []error{}
+	errors = append(errors, deleteServicePortClaim(oldSvc)...)
+	errors = append(errors, addServicePortClaim(newSvc)...)
+	return errors
+}
+
+func handleServicePortClaim(svc *kapi.Service, shouldOpen bool) []error {
+	errors := []error{}
+	if !util.ServiceTypeHasNodePort(svc) && len(svc.Spec.ExternalIPs) == 0 {
+		return errors
+	}
+	for _, port := range svc.Spec.Ports {
+		if util.ServiceTypeHasNodePort(svc) {
+			errors = append(errors, handlePortClaim(port.NodePort, port.Protocol, svc, shouldOpen)...)
+		}
+		if len(svc.Spec.ExternalIPs) > 0 {
+			errors = append(errors, handlePortClaim(port.Port, port.Protocol, svc, shouldOpen)...)
+		}
+	}
+	return errors
+}
+
+func handlePortClaim(portNum int32, protocol kapi.Protocol, svc *kapi.Service, shouldOpen bool) []error {
+	errors := []error{}
+	if err := util.ValidatePort(protocol, portNum); err != nil {
+		errors = append(errors, fmt.Errorf("invalid service port %s, err: %v", svc.Name, err))
+		return errors
+	}
+	if shouldOpen {
+		if err := port.open(portNum, protocol, svc); err != nil {
+			errors = append(errors, err)
+		}
+	} else {
+		if err := port.close(portNum, svc); err != nil {
+			errors = append(errors, err)
+		}
+	}
+	return errors
+}
+
+func (p *portClaimWatcher) open(port int32, protocol kapi.Protocol, svc *kapi.Service) error {
+	klog.V(5).Infof("opening socket for service: %s and port: %v", svc.Name, port)
+	var socket activeSocket
+	var socketError error
+	switch protocol {
+	case kapi.ProtocolTCP:
+		listener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+		if err != nil {
+			socketError = err
+			break
+		}
+		socket = listener
+	case kapi.ProtocolUDP:
+		addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf(":%d", port))
+		if err != nil {
+			socketError = err
+			break
+		}
+		conn, err := net.ListenUDP("udp", addr)
+		if err != nil {
+			socketError = err
+			break
+		}
+		socket = conn
+	case kapi.ProtocolSCTP:
+		// TOOD: implement
+		socketError = fmt.Errorf("unsupported protocol: %q", protocol)
+	default:
+		socketError = fmt.Errorf("unknown protocol %q", protocol)
+	}
+	if socketError != nil {
+		p.emitPortClaimEvent(svc, port, socketError)
+		return socketError
+	}
+	p.activeSockets[port] = socket
```

Comment:
do you need proper locking around this in case there were multiple consecutive calls to open and close the port?

Reply:
I think we need the locking because the factory.go is going away in favor of workqueues and even though we know today service uses a serial informer instead of a queued informer that wont always be the case. Therefore its safer here to add locking on the map (not thread safe) to avoid an issue down the road.

Reply:
As we agreed, I added a lock around all `activeSockets` accesses.